### PR TITLE
Bypass chain-selection subsystem until disputes are enabled.

### PR DESCRIPTION
**Pull request by drahnr**
<del>The `cfg!(feature = "fx")` does not exist yet, but it does not matter for compilation.</del> Uses the `feature = "disputes"` as gate for the relay chain selection rule.

### CODE REVIEW

**Review by eskimor — APPROVED** [review]

**Comment by rphmeier** [comment]
(as discussed, needs to feature-gate disputes lag checking as well)

**Comment by drahnr** [comment]
The check labels bot is wrong.

**Comment by s3krit** [comment]
check_labels should be happy now

File: node/service/src/relay_chain_selection.rs [hunk 1]
```diff
@@ -419,30 +427,35 @@ where
...
+			let lag_disputes = initial_leaf_number.saturating_sub(subchain_number);
+			self.metrics.note_disputes_finality_lag(lag_disputes);
+			lag_disputes
+		} else {
+			lag
+		};
 
 		// 4. Apply the maximum safeguard to the finality lag.
-		if lag > MAX_FINALITY_LAG {
+		if lag < MAX_FINALITY_LAG {
```
Comment (eskimor) [inline node/service/src/relay_chain_selection.rs]: This does not look right.

File: node/service/src/relay_chain_selection.rs [hunk 2]
```diff
@@ -313,14 +332,15 @@ where
 	///
 	/// It will also constrain the chain to only chains which are fully
 	/// approved, and chains which contain no disputes.
-	async fn finality_target(
+	pub async fn finality_target_with_fallback(
```
Comment (rphmeier) [inline node/service/src/relay_chain_selection.rs:335]: Why  is this now pub?

File: node/service/src/relay_chain_selection.rs [hunk 3]
```diff
@@ -304,6 +309,20 @@ where
 		self.block_header(best_leaf)
 	}
 
+	async fn finality_target(
+		&self,
+		_target_hash: Hash,
+		_maybe_max_number: Option<BlockNumber>,
+	) -> Result<Option<Hash>, ConsensusError> {
+		unreachable!("Must call finality_target_with_fallback instead. qed");
```
Comment (rphmeier) [inline node/service/src/relay_chain_selection.rs:317]: This exists in a trait implementation which is called by the GRANDPA code. Wouldn't that unconditionally hit the panic in this case?

File: node/service/src/relay_chain_selection.rs [hunk 4]
```diff
@@ -419,27 +444,32 @@ where
...
-				},
-				std::any::type_name::<Self>(),
-			)
-			.await;
-		let (subchain_number, subchain_head) = rx
-			.await
-			.map_err(Error::OverseerDisconnected)
-			.map_err(|e| ConsensusError::Other(Box::new(e)))?
-			.unwrap_or_else(|| (subchain_number, subchain_head));
+		let lag = if cfg!(feature = "disputes") {
```
Comment (rphmeier) [inline node/service/src/relay_chain_selection.rs:447]: 👍

File: node/service/src/relay_chain_selection.rs (hunk 3 above)
Comment (drahnr) [reply node/service/src/relay_chain_selection.rs:317]: no, we only pass a wrapping type which contains both `selection` and `fallback`. This wrapper will be removed once the overseer proc macro gains the ability to use closures for initialization and avoid the `connected` /`disconnected` states.

Hence the unreachable.

File: node/service/src/relay_chain_selection.rs (hunk 2 above)
Comment (drahnr) [reply node/service/src/relay_chain_selection.rs:335]:
```suggestion
	async fn finality_target_with_fallback(
```

File: node/service/src/relay_chain_selection.rs (hunk 3 above)
Comment (rphmeier) [reply node/service/src/relay_chain_selection.rs:317]: I'd prefer to just remove the unused trait implementation and move the other methods to inherent impl

Comment (drahnr) [reply node/service/src/relay_chain_selection.rs:317]: Yeah, you are correct. The initial idea of just removing `struct RelayChainSelWithFallback` is obsolete now that we need fallback unconditionally. 👍

File: node/service/src/relay_chain_selection.rs [hunk 5]
```diff
@@ -339,6 +344,11 @@ where
 				None => return Ok(Some(target_hash)),
 				Some(best) => best,
 			}
+		} else {
+			match best_leaf {
+				None => return Ok(Some(target_hash)),
+				Some(best_leaf) => best_leaf,
+			}
```
Comment (drahnr) [inline node/service/src/relay_chain_selection.rs:351]: @rphmeier could double check this one? The case where we did not get a best/longest chain in the first place. I think this is the best we can do in that case.

Comment (rphmeier) [reply node/service/src/relay_chain_selection.rs:351]: Yeah, I also agree. We'll need to rethink when we re-enable chain selection

File: node/service/src/tests.rs [hunk 6]
```diff
@@ -1,3 +1,4 @@
+#![allow(dead_code)]
```
Comment (shawntabrizi) [inline node/service/src/tests.rs:1]: maybe create an issue / todo so this doesnt live forever in this file by accident

File: node/service/src/relay_chain_selection.rs [hunk 7]
```diff
@@ -419,27 +429,32 @@ where
...
+			// The the total lag accounting for disputes.
+			let lag_disputes = initial_leaf_number.saturating_sub(subchain_number);
+			self.metrics.note_disputes_finality_lag(lag_disputes);
+			lag_disputes
+		} else {
+			lag
+		};
 
 		// 4. Apply the maximum safeguard to the finality lag.
 		if lag > MAX_FINALITY_LAG {
```
Comment (shawntabrizi) [inline node/service/src/relay_chain_selection.rs:460]: im reviewing out of context of all this code, but the behavior here before and after is different.

This check before was always checking against `lag`, and with the new code, could check against `lag_disputes` with the feature flag.

Was this a bug before? or a bug now?

Comment (rphmeier) [reply node/service/src/relay_chain_selection.rs:460]: Bug before, addressed here. Approval and disputes both apply constraints to the finalizable chain and previously the disputes lag wasn't accounted for in the 'max' check.

File: node/service/src/tests.rs (hunk 6 above)
Comment (rphmeier) [reply node/service/src/tests.rs:1]: @drahnr

Comment (eskimor) [reply node/service/src/tests.rs:1]: Done: https://github.com/paritytech/polkadot/issues/3674

**Pull request merged**